### PR TITLE
Clarify layer ordering

### LINKED ISSUE #276: clarify order of `layers` in image-layout
There is currently no ordering about the ordering of layers in [image-layout](./image-layout.md).
For that matter the bottom-up phrasing in [config](./config.md) can be clarified a bit.

### PR DESCRIPTION
**Pull request by vbatts**
This series of commits touches on the places that ordering is applied/unpacked.
Attempting to have consistent wording.

Fixes #276

### CODE REVIEW

**Comment by vbatts** [comment]
@opencontainers/image-spec-maintainers PTAL (and @aaronlehman too, please :-)

File: config.md [hunk 1]
```diff
@@ -141,12 +143,12 @@ Since the configuration JSON that gets hashed references hashes of each layer in
 
     - **diff_ids** *array*, REQUIRED
 
-       An array of layer content hashes (`DiffIDs`), in order from bottom-most to top-most.
+       An array of layer content hashes (`DiffIDs`), in stack order from bottom-most to top-most (i.e. from diff_ids[0] to diff_ids[len(diff_ids)]).
```
Comment (wking) [inline config.md]: “diff_ids[0] to diff_ids[len(diff_ids)]” → “`diff_ids[0]` to `diff_ids[len(diff_ids)]`”.

File: manifest.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ The first goal is content-addressable images, by supporting an image model where
 The second goal is to allow multi-architecture images, through a "fat manifest" which references image manifests for platform-specific versions of an image.
 The third goal is to be translatable to the [OpenContainers/runtime-spec](https://github.com/opencontainers/runtime-spec)
 
+This section defines the `application/vnd.oci.image.manifest.list.v1+json` and `application/vnd.oci.image.manifest.v1+json` [media type](media-types.md).
```
Comment (wking) [inline manifest.md]: Can we break this up into two files?  Manifest lists and manifests seem as separate as manifests and configs.  And while I like the commits that clearly list the media types being defined, I think that change is orthogonal to the order-clarification.  Maybe it belongs in a separate PR?

File: manifest.md [hunk 3]
```diff
@@ -159,8 +160,8 @@ Unlike the [Manifest List](#manifest-list), which contains information about a s
 
     Each item in the array MUST be a [descriptor](descriptor.md).
     The array MUST have the base image at index 0.
-    Subsequent layers MUST then follow in the order in which they are to be layered on top of each other.
-    The algorithm to create the final unpacked filesystem layout MUST be to first unpack the layer at index 0, then index 1, and so on.
+    Subsequent layers MUST then follow in stack order.
+    The final filesystem layout MUST be from [applying](layer.md#applying) the layers in stack order (i.e. from layers[0] to layers[len(layers)]).
```
Comment (wking) [inline manifest.md]: I still think we need to specify something about the initial state before the layers are applied (#318).  I also [changed “be” to “match” in that PR to allow for implementations like union mounting][1].

Also backticks around the `layers[…]` stuff.

[1]: https://github.com/opencontainers/image-spec/pull/318#issue-177482014

File: image-layout.md [hunk 4]
```diff
@@ -2,7 +2,11 @@
 
 The OCI Image Layout is a slash separated layout of OCI content-addressable blobs and [location-addressable](https://en.wikipedia.org/wiki/Content-addressable_storage#Content-addressed_vs._location-addressed) references (refs).
 This layout MAY be used in a variety of different transport mechanisms: archive formats (e.g. tar, zip), shared filesystem environments (e.g. nfs), or networked file fetching (e.g. http, ftp, rsync).
-Given an image layout a tool can convert a given ref into a runnable OCI Image Format by finding an appropriate manifest from the manifest list, unpacking the [filesystem layers](layer.md) in the correct order, and then converting the image configuration into an OCI Runtime config.json.
+
+Tooling can:
+* convert a given ref into a runnable OCI Image Format by finding an appropriate manifest from the manifest list
+* apply the [filesystem layers](layer.md) per the [manifest layers ordering](manifest.md#image-manifest-property-descriptions)
+* convert the image configuration into an [OCI Runtime config.json](https://github.com/opencontainers/runtime-spec)
```
Comment (wking) [inline image-layout.md]: All of these steps are required to create a runable bundle.  How about:

> Given an image layout and a ref, a tool can create a runnable [OCI Runtime Specification bundle][bundle] by:
>
> 1. Following the ref to find a [manifest](link), possibly via a [manifest list](link),
> 2. Unpacking the [filesystem layers](layer.md) in the correct order, and
> 3. Converting the [image configuration](link) into an [OCI Runtime Specification `config.json`][runtime-config].

[bundle]: https://github.com/opencontainers/runtime-spec/blob/v1.0.0-rc2/bundle.md
[runtime-config]: https://github.com/opencontainers/runtime-spec/blob/v1.0.0-rc2/config.md

File: manifest.md (hunk 3 above)
Comment (philips) [inline manifest.md]: Agreed, but lets leave this for #318

**Comment by vbatts** [comment]
I see that #318 removes that line, but not how it clarifies anything. What
do you want me to change?

**Comment by philips** [comment]
@vbatts don't worry about #318. The language in #318 is "+    The root filesystem MUST match the result of [applying](layer.md#applying) the entries to an empty directory in the listed order."

I think this is a good suggestion for this PR: https://github.com/opencontainers/image-spec/pull/349#discussion_r81191601

**Comment by wking** [comment]
On Thu, Sep 29, 2016 at 02:31:15PM -0700, Vincent Batts wrote:
> I see that #318 removes that line, but not how it clarifies
> anything. What do you want me to change?

#318 forbids seeding the unpack directory location with any content
[1] and changes “be” → “match”, but I'm ok with @philips' suggestion
to keep those in #318 [2].  If we punt those back to #318, the only
change I'd like to see for that line is the backticks [3].

[1]: https://github.com/opencontainers/image-spec/pull/318#issuecomment-248430778
[2]: https://github.com/opencontainers/image-spec/pull/349#discussion_r81234826
[3]: https://github.com/opencontainers/image-spec/pull/349#discussion_r81190383

**Comment by vbatts** [comment]
updated

File: layer.md [hunk 5]
```diff
@@ -1,9 +1,11 @@
 # Image Layer Filesystem Changeset
 
 This document describes how to serialize a filesystem and filesystem changes like removed files into a blob called a layer.
-One or more layers are ordered on top of each other to create a complete filesystem.
+One or more layers changesets are applied on top of each other to create a complete filesystem.
```
Comment (jonboulle) [inline layer.md]: remove "changesets"

File: config.md [hunk 6]
```diff
@@ -141,12 +143,12 @@ Since the configuration JSON that gets hashed references hashes of each layer in
 
     - **diff_ids** *array*, REQUIRED
 
-       An array of layer content hashes (`DiffIDs`), in order from bottom-most to top-most.
+       An array of layer content hashes (`DiffIDs`), in stack order from bottom-most to top-most (i.e. from `diff_ids[0]` to `diff_ids[len(diff_ids)]`).
```
Comment (jonboulle) [inline config.md]: I really don't like introducing the word "stack" because it addles my poor brain

File: image-layout.md [hunk 7]
```diff
@@ -2,7 +2,12 @@
 
 The OCI Image Layout is a slash separated layout of OCI content-addressable blobs and [location-addressable](https://en.wikipedia.org/wiki/Content-addressable_storage#Content-addressed_vs._location-addressed) references (refs).
 This layout MAY be used in a variety of different transport mechanisms: archive formats (e.g. tar, zip), shared filesystem environments (e.g. nfs), or networked file fetching (e.g. http, ftp, rsync).
-Given an image layout a tool can convert a given ref into a runnable OCI Image Format by finding an appropriate manifest from the manifest list, unpacking the [filesystem layers](layer.md) in the correct order, and then converting the image configuration into an OCI Runtime config.json.
+
+Given an image layout and a ref, a tool can create an [OCI Runtime Specification bundle](https://github.com/opencontainers/runtime-spec/blob/v1.0.0-rc2/bundle.md) by:
+
+* Following the ref to find a [manifest](manifest.md#image-manifest), possibly via a [manifest list](manifest.md#manifest-list)
+* Apply the filesystem layers in the specified order
```
Comment (jonboulle) [inline image-layout.md]: Applying

File: config.md [hunk 8]
```diff
@@ -3,6 +3,8 @@
 An *Image* is an ordered collection of root filesystem changes and the corresponding execution parameters for use within a container runtime.
 This specification outlines the JSON format describing images for use with a container runtime and execution tool and its relationship to filesystem changesets, described in [Layers](layer.md).
 
+This section defines the `application/vnd.oci.image.config.v1+json` [media type](media-types.md).
```
Comment (jonboulle) [inline config.md]: I appreciate these but would also say a different PR..

Comment (vbatts) [inline config.md]: k

File: config.md (hunk 6 above)
Comment (vbatts) [inline config.md]: k. I'll just switch from the whole bottom to top, since we're not pushing and popping this list anyways.

File: image-layout.md (hunk 7 above)
Comment (vbatts) [inline image-layout.md]: k

File: layer.md (hunk 5 above)
Comment (vbatts) [inline layer.md]: k

**Comment by vbatts** [comment]
updated PTAL

**Comment by jonboulle** [comment]
lgtm

File: image-layout.md [hunk 9]
```diff
@@ -2,7 +2,12 @@
 
 The OCI Image Layout is a slash separated layout of OCI content-addressable blobs and [location-addressable](https://en.wikipedia.org/wiki/Content-addressable_storage#Content-addressed_vs._location-addressed) references (refs).
 This layout MAY be used in a variety of different transport mechanisms: archive formats (e.g. tar, zip), shared filesystem environments (e.g. nfs), or networked file fetching (e.g. http, ftp, rsync).
-Given an image layout a tool can convert a given ref into a runnable OCI Image Format by finding an appropriate manifest from the manifest list, unpacking the [filesystem layers](layer.md) in the correct order, and then converting the image configuration into an OCI Runtime config.json.
+
+Given an image layout and a ref, a tool can create an [OCI Runtime Specification bundle](https://github.com/opencontainers/runtime-spec/blob/v1.0.0-rc2/bundle.md) by:
+
+* Following the ref to find a [manifest](manifest.md#image-manifest), possibly via a [manifest list](manifest.md#manifest-list)
+* Applying the filesystem layers in the specified order
```
Comment (wking) [inline image-layout.md]: Make “Applying the filesystem layers” a link to `layers.go#applying`?

File: image-layout.md [hunk 10]
```diff
@@ -2,7 +2,12 @@
...
+* Converting the [image configuration](config.md) into an [OCI Runtime Specification config.json](https://github.com/opencontainers/runtime-spec/blob/v1.0.0-rc2/config.md)
```
Comment (wking) [inline image-layout.md]: Backticks around `config.json`?

**Comment by philips** [comment]
LGTM
